phoenix-core/src/main/java/org/apache/phoenix/compile/ServerBuildIndexCompiler.java

```diff
@@ -0,0 +1,133 @@
+            final PTable dataTable = tableRef.getTable();
+            if (index.getIndexType() == PTable.IndexType.GLOBAL &&  dataTable.isTransactional()) {
+                throw new IllegalArgumentException(
+                        "ServerBuildIndexCompiler does not support global indexes on transactional tables");
+            }
+            IndexMaintainer.serialize(dataTable, ptr, Collections.singletonList(index), plan.getContext().getConnection());
+            // Set the scan attributes that UngroupedAggregateRegionObserver will switch on.
+            // For local indexes, the BaseScannerRegionObserver.LOCAL_INDEX_BUILD_PROTO attribute, and
+            // for global indexes PhoenixIndexCodec.INDEX_PROTO_MD attribute is set to the serialized form of index
+            // metadata to build index rows from data table rows. For global indexes, we also need to set (1) the
+            // BaseScannerRegionObserver.REBUILD_INDEXES attribute in order to signal UngroupedAggregateRegionObserver
+            // that this scan is for building global indexes and (2) the MetaDataProtocol.PHOENIX_VERSION attribute
+            // that will be passed as a mutation attribute for the scanned mutations that will be applied on
+            // the index table possibly remotely
+            if (index.getIndexType() == PTable.IndexType.LOCAL) {
+                scan.setAttribute(BaseScannerRegionObserver.LOCAL_INDEX_BUILD_PROTO, ByteUtil.copyKeyBytesIfNecessary(ptr));
+            } else {
+                scan.setAttribute(PhoenixIndexCodec.INDEX_PROTO_MD, ByteUtil.copyKeyBytesIfNecessary(ptr));
+                scan.setAttribute(BaseScannerRegionObserver.REBUILD_INDEXES, TRUE_BYTES);
+                ScanUtil.setClientVersion(scan, MetaDataProtocol.PHOENIX_VERSION);
+            }
+            // By default, we'd use a FirstKeyOnly filter as nothing else needs to be projected for count(*).
+            // However, in this case, we need to project all of the data columns that contribute to the index.
+            IndexMaintainer indexMaintainer = index.getIndexMaintainer(dataTable, connection);
+            for (ColumnReference columnRef : indexMaintainer.getAllColumns()) {
+                if (index.getImmutableStorageScheme() == PTable.ImmutableStorageScheme.SINGLE_CELL_ARRAY_WITH_OFFSETS) {
+                    scan.addFamily(columnRef.getFamily());
+                } else {
+                    scan.addColumn(columnRef.getFamily(), columnRef.getQualifier());
+                }
+            }
+
+            if (dataTable.isTransactional()) {
+                scan.setAttribute(BaseScannerRegionObserver.TX_STATE, connection.getMutationState().encodeTransaction());
+            }
+
+            // Go through MutationPlan abstraction so that we can create local indexes
+            // with a connectionless connection (which makes testing easier).
+            return new BaseMutationPlan(plan.getContext(), PhoenixStatement.Operation.UPSERT) {
```

Comment:
nit: I'm trying to get rid of non-trivial anonymous classes in Phoenix, so let's not add more :-)

Reply:
I will convert it to an inner class